src/reducers/viewconfig.ts

```diff
@@ -128,140 +176,166 @@ const getAllDescendantIds = (state: ViewConfigReducer, nodeId: number) =>
         case 'ADD_VIEW_CONFIG_NODE': {
-            const id = lastId(items) + 1;
+            const newId = childNodes.length > 0 ? [childNodes[childNodes.length - 1].id] : [];
             return {
-                ...action.payload.component,
-                id,
-                childIds: [],
-                name: createName(action.payload.component.type, id)
+                ...(state as NormalizedComponentConfig),
+                childIds: [...(state as NormalizedComponentConfig).childIds, ...newId]
             };
         }
-        case 'MODIFY_VIEW_CONFIG_NODE': // TODO: Make sure it get's normalized in case of total component switch. It doesn't break, but it's not nice either
+        case 'MODIFY_VIEW_CONFIG_NODE':
             return {
                 ...action.payload.component,
-                id: state!.id,
-                childIds: state!.childIds,
-                name: createName(action.payload.component.type, state!.id)
+                id: (state as NormalizedComponentConfig).id,
+                childIds: (state as NormalizedComponentConfig).childIds,
+                name: createName(action.payload.component.type, (state as NormalizedComponentConfig).id)
             };
-        case 'ADD_VIEW_CONFIG_CHILD':
-        case 'DELETE_VIEW_CONFIG_CHILD':
         case 'SORT_VIEW_CONFIG_CHILD':
             return {
-                ...state!,
-                childIds: childIds(state!.childIds, action)
+                ...(state as NormalizedComponentConfig),
+                childIds: arrayMove(
+                    (state as NormalizedComponentConfig).childIds,
+                    action.payload.oldIndex,
+                    action.payload.newIndex
+                )
+            };
+        case 'SWITCH_VIEW_CONFIG_NODE': {
+            const newNodeList = normalizeTree([action.payload.component], action.payload.collectionId);
+            const newNode = newNodeList[newNodeList.length - 2];
+            return {
+                ...newNode,
+                id: (state as NormalizedComponentConfig).id,
+                childIds: childNodes.map(childNode => childNode.id),
+                name: createName(action.payload.component.type, (state as NormalizedComponentConfig).id)
             };
+        }
         default:
-            return state!; // todo don't unwrap optional???
+            return state as NormalizedComponentConfig; // todo don't unwrap optional???
     }
 };
 
 export default (state = initialState, action: Action): ViewConfigReducer => {
     switch (action.type) {
         case 'MODIFY_VIEW_CONFIG_NODE':
         case 'SORT_VIEW_CONFIG_CHILD':
-        case 'DELETE_VIEW_CONFIG_CHILD':
-        case 'ADD_VIEW_CONFIG_CHILD': {
-            const nodeId = action.payload.nodeId;
-            const nodeIndex = state.findIndex(item => item.id === nodeId);
-            const nextNode = node(state[nodeIndex], action, state);
+        case 'ADD_VIEW_CONFIG_NODE':
+        case 'SWITCH_VIEW_CONFIG_NODE': {
```

Comment:
je had me uitgelegd wat switchen dee, maar ik ben het alweer vergeten, is er niet een betere term hier voor te bedenken?